### PR TITLE
Added AD versions of TwoPhaseFluidProperties interfaces

### PR DESCRIPTION
**Pull request by joshuahansel**

### CODE REVIEW

**Comment by joshuahansel** [comment]
As discussed offline, it is probably best to use a non-test-object instead of `TestTwoPhaseFluidProperties` since this should not be used in unit tests, only regression tests.

**Comment by joshuahansel** [comment]
THM-based apps will fail to compile due to #15288, so those targets will need to be invalidated after the THM fixes get to its master branch.

File: modules/fluid_properties/include/userobjects/TwoPhaseFluidProperties.h [hunk 1]
```diff
@@ -44,13 +47,15 @@ class TwoPhaseFluidProperties : public FluidProperties
    * @param[in] p  pressure
    */
   virtual Real T_sat(Real p) const = 0;
+  virtual DualReal T_sat(DualReal p) const;
```
Comment (rwcarlsen) [inline modules/fluid_properties/include/userobjects/TwoPhaseFluidProperties.h:50]: `const DualReal &` here and elsewhere.

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
Like the unit tests.  Just the const ref on func args.

**Comment by joshuahansel** [comment]
@rwcarlsen I've failed quite a few times due to those GCC overload-virtual errors (see "Modules debug" failure). I put those pragmas to ignore them in all of the 2-phase header files, but it seems that is not enough. Does it ever need to go in the source file? From the failure, it looks like it.

**Comment by rwcarlsen** [comment]
Not sure - I've never needed to do that before.  Do the pragmas in the C file fix the error?

**Comment by joshuahansel** [comment]
> Do the pragmas in the C file fix the error?

Let's find out with my next push. I bet it will work - I just wanted to understand those errors a bit to understand when exactly it's needed.

**Comment by rwcarlsen** [comment]
I think it needs to go in wherever the implementation lives.  The hidden virtual funcs' implementation (the DualReal versions) for single phase were all in the header file (using a macro) - so it does seem to make sense that you'd need it in the C file since that is where the func bodies are.

**Review by rwcarlsen — APPROVED** [review]

**Pull request merged**